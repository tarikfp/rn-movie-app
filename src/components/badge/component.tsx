import * as React from "react";
import { styled } from "~theme";

type Props = {
  label: string;
};

export default function Badge({ label }: Props) {
  return (
    <BadgeContainer>
      <BadgeText>{label}</BadgeText>
    </BadgeContainer>
  );
}

const BadgeContainer = styled.View`
  border-radius: 16px;
  height: 24px;
  min-width: 105px;
  margin-right: 6px;
  margin-top: 6px;
  justify-content: center;
  align-items: center;
  background-color: ${({ theme: { colors } }) => colors.primary};
`;

const BadgeText = styled.Text`
  background-color: ${({ theme: { colors } }) => colors.primary};
  color: ${({ theme: { colors } }) => colors.paper};
  font-weight: ${({ theme: { fonts } }) => fonts.p3.fontWeight};
  font-size: ${({ theme: { fonts } }) => fonts.p3.fontSize}px;
  font-family: ${({ theme: { fonts } }) => fonts.p3.fontFamily};
`;
