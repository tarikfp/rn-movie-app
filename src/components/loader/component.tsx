import * as React from "react";
import { ActivityIndicatorProps, Platform } from "react-native";
import { ms } from "react-native-size-matters";
import styled from "~theme/styled-theme";

type Props = Pick<ActivityIndicatorProps, "size" | "color">;

export default function Loader({ color, size }: Props) {
  return (
    <StyledActivityIndicator
      animating
      color={(color as string) ?? ""}
      {...Platform.select({
        ios: { size: size ?? "large" },
        android: { size: size ?? ms(26) },
      })}
    />
  );
}

const StyledActivityIndicator = styled.ActivityIndicator.attrs(
  ({ theme: { colors } }) => ({
    color: colors.primary,
  }),
)``;
