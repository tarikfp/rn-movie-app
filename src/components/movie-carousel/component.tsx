import { getWindowWidth } from "@tarikfp/react-native-utils";
import React from "react";
import { ListRenderItemInfo } from "react-native";
import { ms } from "react-native-size-matters";
import Carousel from "react-native-snap-carousel";
import styled from "styled-components/native";
import { MovieTypes } from "~api/movie";

type Props = {
  carouselKey: string;
  // programatically auto play sliders if needed
  autoPlay?: boolean;
  genreName: string;
  order: number;
  data: Array<MovieTypes.Result>;
  renderItem: (
    item: MovieTypes.Result,
    index: number,
    onPressCarouselItem: () => void,
  ) => React.ReactElement;
};

const SLIDER_AUTO_PLAY_DELAY = 2000;
const SLIDER_AUTO_PLAY_INTERVAL = 1000;

export default function MovieCarousel({
  carouselKey,
  data,
  order,
  autoPlay = false,
  genreName,
  renderItem,
}: Props) {
  const carouselRef = React.useRef<Carousel<MovieTypes.Result> | null>(null);

  const onPressCarouselItem = React.useCallback((index: number) => {
    carouselRef.current?.snapToItem(index);
  }, []);

  const memoizedRenderItem = React.useCallback(
    ({ index, item }: ListRenderItemInfo<MovieTypes.Result>) => {
      return renderItem(item, index, () => onPressCarouselItem(index));
    },
    [onPressCarouselItem, renderItem],
  );

  return (
    <CarouselContainer>
      <CategoryText>{genreName}</CategoryText>
      <Carousel
        key={carouselKey}
        ref={carouselRef}
        autoplay={autoPlay}
        loop={false}
        enableSnap
        autoplayInterval={SLIDER_AUTO_PLAY_INTERVAL * (order + 1)}
        autoplayDelay={SLIDER_AUTO_PLAY_DELAY * (order + 1)}
        layout="default"
        sliderWidth={getWindowWidth(100)}
        itemWidth={ms(180)}
        firstItem={data.length / 2}
        data={data}
        renderItem={memoizedRenderItem}
      />
    </CarouselContainer>
  );
}

const CategoryText = styled.Text`
  color: ${({ theme: { colors } }) => colors.textDark};
  font-size: ${({ theme: { fonts } }) => fonts.h2.fontSize}px;
  font-weight: ${({ theme: { fonts } }) => fonts.h2.fontWeight};
  font-family: ${({ theme: { fonts } }) => fonts.h2.fontFamily};
  margin-bottom: 16px;
`;

const CarouselContainer = styled.View`
  justify-content: space-around;
  align-items: center;
`;
