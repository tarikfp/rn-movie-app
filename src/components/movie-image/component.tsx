import * as React from "react";
import FastImage, { FastImageProps } from "react-native-fast-image";
import { styled } from "~theme";

type Props = FastImageProps & {
  height: number;
  width: number;
};

function MovieImage(props: Props) {
  return (
    <Shadow>
      <FastImage {...props} />
    </Shadow>
  );
}

export const StyledMovieImage = styled(MovieImage).attrs<Props, Props>(
  (props) => ({
    resizeMode: FastImage.resizeMode.cover,
    height: props.height,
    width: props.width,
  }),
)<Props>`
  width: ${({ width }) => `${width}px`};
  height: ${({ height }) => `${height}px`};
  shadow-opacity: 0.75;
  shadow-radius: 5px;
  border-radius: 12px;
  shadow-offset: 0px 0px;
  elevation: 1;
`;

const Shadow = styled.View`
  shadow-color: ${({ theme: { colors } }) => colors.black};
  shadow-opacity: 0.8;
  shadow-offset: 1px 3px;
  shadow-radius: 2px;
  elevation: 2;
`;
