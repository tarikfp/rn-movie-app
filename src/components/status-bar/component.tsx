import React from "react";
import { StatusBar, View } from "react-native";
import { useSafeAreaInsets } from "react-native-safe-area-context";
import { theme } from "~theme";

type Props = {
  backgroundColor: string;
};

export default function AppStatusBar({ backgroundColor }: Props) {
  const { top } = useSafeAreaInsets();
  return (
    <View
      style={{
        backgroundColor: backgroundColor ?? theme.colors.transparent,
        height: top,
      }}>
      <StatusBar
        barStyle="light-content"
        backgroundColor={theme.colors.transparent}
        translucent
      />
    </View>
  );
}
