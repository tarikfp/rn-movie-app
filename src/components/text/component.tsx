import * as React from "React";
import { DefaultTheme } from "styled-components/native";

export type FontSize =
  | "smallest"
  | "small"
  | "smallMedium"
  | "medium"
  | "large"
  | "xLarge";

const getFont = (
  fontSize: FontSize,
  theme: DefaultTheme,
  fontWeight?: string,
): string => {
  switch (fontSize) {
    case "smallest":
      return `${fontWeight || theme.fonts.p2.fontWeight} ${
        theme.fonts.p2.fontSize
      }px ${theme.fonts.p2.fontFamily}`;
    case "small":
      return `${fontWeight || theme.fonts.p1.fontWeight} ${
        theme.fonts.p1.fontSize
      }px ${theme.fonts.p1.fontFamily}`;
    case "smallMedium":
      return `${fontWeight || theme.fonts.h4.fontWeight} ${
        theme.fonts.h4.fontSize
      }px ${theme.fonts.h4.fontFamily}`;
    case "medium":
      return `${fontWeight || theme.fonts.h3.fontWeight} ${
        theme.fonts.h3.fontSize
      }px ${theme.fonts.h3.fontFamily}`;
    case "large":
      return `${fontWeight || theme.fonts.h2.fontWeight} ${
        theme.fonts.h2.fontSize
      }px ${theme.fonts.h2.fontFamily}`;
    case "xLarge":
      return `${fontWeight || theme.fonts.h1.fontWeight} ${
        theme.fonts.h1.fontSize
      }px ${theme.fonts.h1.fontFamily}`;
    default:
      return `${theme.fonts.h3.fontWeight} ${theme.fonts.h3.fontSize}px ${theme.fonts.h3.fontFamily}`;
  }
};

export default function AppText(
  props: React.PropsWithChildren<GenericTextProps & TextProps>,
) {
  return <StyledText testID={AppTextTestIds.Root} {...props} />;
}

const StyledText = styled.Text<GenericTextProps>`
  font: ${(props) => getFont(props.fontSize, props.theme, props.fontWeight)};
  text-align: ${(props) => (props.textAlign ? props.textAlign : "center")};
  text-transform: ${(props) =>
    props.textTransform ? props.textTransform : "none"};
  color: ${(props) => (props.color ? props.color : props.theme.colors.text)};
  margin-top: ${(props) => (props.marginTop ? props.marginTop : 0)}px;
  margin-bottom: ${(props) => (props.marginBottom ? props.marginBottom : 0)}px;
  max-width: ${(props) => (props.maxWidth ? props.maxWidth : 300)}px;
  margin-right: ${(props) => (props.marginRight ? props.marginRight : 0)}px;
  margin-left: ${(props) => (props.marginLeft ? props.marginLeft : 0)}px;
`;
