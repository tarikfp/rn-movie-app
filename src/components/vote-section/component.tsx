import { Icon } from "@tarikfp/react-native-utils";
import * as React from "react";
import { styled } from "~theme";

type Props = {
  voteAverage: number;
  voteCount: number;
};
export default function VoteSection({ voteAverage, voteCount }: Props) {
  return (
    <VoteContainer>
      <Icon disabled name="star" type="MaterialCommunityIcons" color="orange" />
      <VoteText>
        {" "}
        {voteAverage.toFixed(2)}/10 {"  "}({voteCount} reviews)
      </VoteText>
    </VoteContainer>
  );
}

const VoteText = styled.Text`
  color: ${({ theme: { colors } }) => colors.textDark};
  font-size: ${({ theme: { fonts } }) => fonts.p3.fontSize}px;
  font-weight: ${({ theme: { fonts } }) => fonts.p3.fontWeight};
  font-family: ${({ theme: { fonts } }) => fonts.p3.fontFamily};
`;

const VoteContainer = styled.View`
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  margin-top: 12px;
`;
