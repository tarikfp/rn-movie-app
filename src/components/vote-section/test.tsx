import React from "react";
import { TestUtils } from "~utils";
import { default as VoteSection } from "./component";

const props: React.ComponentProps<typeof VoteSection> = {
  voteAverage: 6.54,
  voteCount: 2423,
};

describe("Tests vote section component", () => {
  it("renders vote section correctly", () => {
    const tree = TestUtils.createWithWrappers(<VoteSection {...props} />, {
      hasTheme: true,
      hasSafeAreaProvider: true,
    });

    expect(tree).toMatchSnapshot();
  });
});
