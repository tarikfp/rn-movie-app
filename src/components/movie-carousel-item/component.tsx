import * as React from "react";
import { Pressable } from "react-native";
import { ms } from "react-native-size-matters";
import { MovieTypes } from "~api/movie";
import { styled } from "~theme";
import { getImageSourceUri } from "~utils";
import { MovieImage } from "../movie-image";
import { VoteSection } from "../vote-section";

export default function MovieCarouselItem(
  props: MovieTypes.MovieListItem & { onPress: () => void },
) {
  return (
    <Pressable onPress={props.onPress}>
      <MovieImage
        height={ms(210)}
        width={ms(180)}
        source={{
          uri: getImageSourceUri(props.backdrop_path),
        }}
      />
      <TitleText>{props.original_title}</TitleText>

      <VoteSection
        voteAverage={props.vote_average}
        voteCount={props.vote_count}
      />
    </Pressable>
  );
}

const TitleText = styled.Text`
  color: ${({ theme: { colors } }) => colors.textDark};
  font-size: ${({ theme: { fonts } }) => fonts.p1.fontSize}px;
  font-weight: ${({ theme: { fonts } }) => fonts.p1.fontWeight};
  font-family: ${({ theme: { fonts } }) => fonts.p1.fontFamily};
  margin-top: 12px;
`;
