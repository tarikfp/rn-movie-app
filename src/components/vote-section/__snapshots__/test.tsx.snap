// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests vote section component renders vote section correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
/>
`;
